fe/fe-core/src/main/java/com/starrocks/metric/MetricRepo.java

```diff
@@ -526,6 +545,76 @@ private static void collectTableMetrics(MetricVisitor visitor, boolean minifyTab
+
+        List<RoutineLoadJob> kafkaJobs = jobs.stream().filter(job -> (job instanceof KafkaRoutineLoadJob)).collect(
+                Collectors.toList());
+
+        if (kafkaJobs.size() <= 0) {
+            return;
+        }
+
+        // get all partitions offset in a batch api
+        List<PKafkaOffsetProxyRequest> requests = new ArrayList<>();
+        for (RoutineLoadJob job : kafkaJobs) {
+            KafkaRoutineLoadJob kJob = (KafkaRoutineLoadJob) job;
+            try {
+                kJob.convertCustomProperties(false);
+            } catch (DdlException e) {
+                LOG.warn("convert custom properties failed", e);
+                return;
+            }
+            PKafkaOffsetProxyRequest offsetProxyRequest = new PKafkaOffsetProxyRequest();
+            offsetProxyRequest.kafkaInfo = KafkaUtil.genPKafkaLoadInfo(kJob.getBrokerList(), kJob.getTopic(),
+                    ImmutableMap.copyOf(kJob.getConvertedCustomProperties()));
+            List<Integer> partitionIds = new ArrayList<>(
+                    ((KafkaProgress) kJob.getProgress()).getPartitionIdToOffset().keySet());
+            offsetProxyRequest.partitionIds = partitionIds;
+            requests.add(offsetProxyRequest);
+        }
+        List<PKafkaOffsetProxyResult> offsetProxyResults;
+        try {
+            offsetProxyResults = KafkaUtil.getBatchOffsets(requests);
+        } catch (UserException e) {
+            LOG.warn("get batch offsets failed", e);
+            return;
+        }
+
+        for (int i = 0; i < kafkaJobs.size(); i++) {
+            KafkaRoutineLoadJob kJob = (KafkaRoutineLoadJob) kafkaJobs.get(i);
+            ImmutableMap<Integer, Long> partitionIdToProgress =
+                    ((KafkaProgress) kJob.getProgress()).getPartitionIdToOffset();
+
+            // offset of partitionIds[i] is beginningOffsets[i] and latestOffsets[i]
+            List<Integer> partitionIds = offsetProxyResults.get(i).partitionIds;
+            List<Long> beginningOffsets = offsetProxyResults.get(i).beginningOffsets;
+            List<Long> latestOffsets = offsetProxyResults.get(i).latestOffsets;
+
+            long maxLag = Long.MIN_VALUE;
+            for (int j = 0; j < partitionIds.size(); j++) {
+                int partitionId = partitionIds.get(j);
+                if (!partitionIdToProgress.containsKey(partitionId)) {
+                    continue;
+                }
+                long progress = partitionIdToProgress.get(partitionId);
+                if (progress == KafkaProgress.OFFSET_BEGINNING_VAL) {
+                    progress = beginningOffsets.get(j);
+                }
+
+                maxLag = Math.max(latestOffsets.get(j) - progress, maxLag);
+            }
+            if (maxLag >= Config.min_routine_load_lag_for_metrics) {
+                GaugeMetricImpl<Long> metric =
+                        new GaugeMetricImpl<>("routine_load_lag", MetricUnit.NOUNIT, "routine load kafka lag");
```

Comment:
The lag is for the max partition, not all the partitions of the job.
Better to use a more clear name.

Reply:
done